src/v/cluster/controller_stm.cc

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/controller_stm.h"
+
+#include "bytes/iostream.h"
+#include "cluster/controller_snapshot.h"
+#include "cluster/logger.h"
+#include "cluster/members_manager.h"
+#include "vlog.h"
+
+namespace cluster {
+
+ss::future<> controller_stm::on_batch_applied() {
+    if (!_feature_table.is_active(features::feature::controller_snapshots)) {
+        co_return;
+    }
+    if (_gate.is_closed()) {
+        co_return;
+    }
+
+    if (
+      get_last_applied_offset() > _raft->last_snapshot_index()
+      && !_snapshot_debounce_timer.armed()) {
+        _snapshot_debounce_timer.arm(_snapshot_max_age());
+    };
```

Comment:
are there extreme cases where there were multiple background fibers running `maybe_write_snapshot` (started via the timer), and if so, is that problematic or handled in `maybe_write_snapshot`?

Reply:
Should be okay, first, snapshotting is protected by a mutex in `mux_state_machine`, and second, even if they queue up on the mutex, there is another offset check in `maybe_write_snapshot`, so all attempts but the first will be no-ops.

Reply:
thanks. i didn't recognize the that a mutex existed. sgtm